### PR TITLE
[Refact] Get rid of PyQTimeDependentEvolution

### PR DESCRIPTION
**Pull request by vytautas-a**
Substitute `PyQTimeDependentEvolution` class with proper call to `pyqtorch`'s time-dependent functionality.

### CODE REVIEW

**Comment by vytautas-a** [comment]
This refactoring disables the embedding function call for `pyqtorch` backend, thus currently only `HamEvo` block works properly. I suggest we could completely get rid of the embedding call when dealing with `pyqtorch`. This can be achieved by utilizing the new `sympy` -> `ConcretizedCallable` converter.

**Comment by Roland-djee** [comment]
Hey @vytautas-a can I mark this PR as ready ?

**Comment by vytautas-a** [comment]
@Roland-djee, well this PR here is currently more in the proposal stage. It depends also on another PR that I'll open in `pyqtorch`.

File: qadence/backends/pyqtorch/convert_ops.py [hunk 1]
```diff
@@ -98,6 +91,53 @@ def extract_parameter(block: ScaleBlock | ParametricBlock, config: Configuration
...
+    # base case - independent argument
+    if len(expr.args) == 0:
+        res = (
+            float(expr)
+            if str(expr).replace(".", "", 1).replace("-", "", 1).isdigit()
+            else str(expr)
+        )
+        if isinstance(res, str):
+            if "/" in res:
+                # found a rational
```
Comment (Roland-djee) [inline qadence/backends/pyqtorch/convert_ops.py:113]:
```suggestion
            if "/" in res:  # Found a rational
```

File: qadence/backends/pyqtorch/convert_ops.py [hunk 2]
```diff
@@ -98,6 +91,53 @@ def extract_parameter(block: ScaleBlock | ParametricBlock, config: Configuration
...
+    Returns:
+        ConcretizedCallable: expression encoded as ConcretizedCallable
+    """
+
+    # base case - independent argument
+    if len(expr.args) == 0:
+        res = (
+            float(expr)
+            if str(expr).replace(".", "", 1).replace("-", "", 1).isdigit()
+            else str(expr)
```
Comment (Roland-djee) [inline qadence/backends/pyqtorch/convert_ops.py:109]: Would it make sense not to convert to a `str` here and use Sympy functionalities later ? Ie. when checking for rationality ?

File: qadence/backends/pyqtorch/convert_ops.py [hunk 3]
```diff
@@ -98,6 +91,53 @@ def extract_parameter(block: ScaleBlock | ParametricBlock, config: Configuration
...
+        if isinstance(res, str):
+            if "/" in res:
+                # found a rational
+                res = float(sympy.Rational(res).evalf())
+            if "fix" in res:  # type: ignore [operator]
+                # found a fixed parameter - convert to float
+                res = torch.as_tensor(float(res.split("_")[1]))  # type: ignore [attr-defined]
+        return res
+
+    # iterate through current function arguments
```
Comment (Roland-djee) [inline qadence/backends/pyqtorch/convert_ops.py:120]:
```suggestion
    # Recursively iterate through current function arguments
```

File: qadence/backends/pyqtorch/convert_ops.py [hunk 4]
```diff
@@ -98,6 +91,53 @@ def extract_parameter(block: ScaleBlock | ParametricBlock, config: Configuration
...
+    all_results = []
+    for arg in expr.args:
+        res = sympy_to_pyq(arg)
+        all_results.append(res)
+
+    # deal with multi-argument (>2) sympy functions: converting to nested
+    # ConcretizedCallable objects
+    if len(all_results) > 2:
+
+        def fn(x: str | ConcretizedCallable, y: str | ConcretizedCallable) -> Callable:
```
Comment (Roland-djee) [inline qadence/backends/pyqtorch/convert_ops.py:130]:
```suggestion
        def convert(x: str | ConcretizedCallable, y: str | ConcretizedCallable) -> Callable:
```

File: qadence/backends/pyqtorch/convert_ops.py [hunk 5]
```diff
@@ -98,6 +91,53 @@ def extract_parameter(block: ScaleBlock | ParametricBlock, config: Configuration
     return config.get_param_name(block)[0]
 
 
+def sympy_to_pyq(expr: sympy.Expr) -> ConcretizedCallable:
+    """Convert sympy expression to pyqtorch ConcretizedCallable object.
```
Comment (Roland-djee) [inline qadence/backends/pyqtorch/convert_ops.py:95]:
```suggestion
    """Convert sympy expression to a possibly nested pyqtorch ConcretizedCallable object.
```

File: qadence/backends/pyqtorch/convert_ops.py [hunk 6]
```diff
@@ -98,6 +91,53 @@ def extract_parameter(block: ScaleBlock | ParametricBlock, config: Configuration
...
+    # deal with multi-argument (>2) sympy functions: converting to nested
+    # ConcretizedCallable objects
+    if len(all_results) > 2:
+
+        def fn(x: str | ConcretizedCallable, y: str | ConcretizedCallable) -> Callable:
+            return partial(ConcretizedCallable, call_name=SYMPY_TO_PYQ_MAPPING[expr.func])(  # type: ignore [no-any-return]
+                abstract_args=[x, y]
+            )
+
+        cc = reduce(fn, all_results)
```
Comment (Roland-djee) [inline qadence/backends/pyqtorch/convert_ops.py:135]:
```suggestion
        concretized_callables = reduce(fn, all_results)
```

File: qadence/execution.py [hunk 7]
```diff
@@ -76,10 +76,14 @@ def _(
         diff_mode = DiffMode.AD
     bknd = backend_factory(backend, diff_mode=diff_mode, configuration=configuration)
     conv = bknd.convert(circuit)
+    if backend == BackendName.PYQTORCH:
+        vals = values
```
Comment (Roland-djee) [inline qadence/execution.py:80]:
```suggestion
        embedded_values = values
```

**Review by Roland-djee — COMMENTED** [review]
Thanks a lot @vytautas-a looks nice and clean. Minor renaming suggestions and comments. Will approve once addressed.

File: qadence/backends/pyqtorch/convert_ops.py [hunk 8]
```diff
@@ -114,36 +157,47 @@ def convert_block(
 
     if isinstance(block, ScaleBlock):
         scaled_ops = convert_block(block.block, n_qubits, config)
-        scale = extract_parameter(block, config)
-        return [pyq.Scale(pyq.Sequence(scaled_ops), scale)]
+        scale = extract_parameter(block, config=config)
+
+        # replace underscore by dot when underscore is between two numbers in string
+        if isinstance(scale, str):
+            scale = re.sub(r"(?<=\d)_(?=\d)", ".", scale)
```
[the remark this answers is not in the document]
Comment (vytautas-a) [reply qadence/backends/pyqtorch/convert_ops.py:164]: This substitution is required due to the method how `qadence` stringifies `sympy` expressions. Your corner case is interesting - I have to test it, but it would seem that current implementation would substitute a dot in this variable. And that is not desirable. The reason for this regular expression is to convert floats from form `0_1234` to `0.1234` that is understood by `parse_expr` method of `sympy`. But yes, it has to be updated to handle your case

[the remark this answers is not in the document]
Comment (vytautas-a) [reply qadence/backends/pyqtorch/convert_ops.py]: fixed

File: tests/engines/test_torch.py [hunk 9]
```diff
@@ -100,7 +100,7 @@ def test_embeddings() -> None:
         "orig_param_values", {}
     )  # TODO: remove this when embedding system is updated
 
-    assert len(list(low_level_params.keys())) == 9
+    assert len(list(low_level_params.keys())) == 9 + len(inputs)
```
Comment (jpmoutinho) [inline tests/engines/test_torch.py:103]: Are you sure this is the right fix?

The line above has a comment to be removed "when embedding system is updated". Could it be that removing that will make the `assert` pass without the extra manual addition?

File: tests/qadence/test_quantum_model.py [hunk 10]
```diff
@@ -52,7 +52,9 @@ def test_quantum_model_parameters(parametric_circuit: QuantumCircuit) -> None:
     embedded_params_ad.pop(
         "orig_param_values", {}
     )  # TODO: remove this when embedding system is updated
-    assert len(embedded_params_ad) == 5
+    assert (
+        len(embedded_params_ad) == 5 + 1
+    )  # adding one because original param x is included for PYQ + AD
```
Comment (jpmoutinho) [inline tests/qadence/test_quantum_model.py:57]: Same as above

File: tests/qadence/test_quantum_model.py [hunk 11]
```diff
@@ -71,7 +73,7 @@ def test_quantum_model_duplicate_expr(duplicate_expression_circuit: QuantumCircu
     embedded_params_ad.pop(
         "orig_param_values", {}
     )  # TODO: remove this when embedding system is updated
-    assert len(embedded_params_ad) == 2
+    assert len(embedded_params_ad) == 2 + 4  # adding 4 because all original params are included
```
Comment (jpmoutinho) [inline tests/qadence/test_quantum_model.py:76]: Same as above

**Review by jpmoutinho — COMMENTED** [review]
Thanks @vytautas-a :)

Few minor things and then it should be good to go.

One thing, I see that the docs for the time-dependent HamEvo reads:

```
# create parameterized HamEvo block
hamevo = HamEvo(generator_td, 0.0, duration=duration)
```

Is passing `0.0` correct there? I thought the idea was to pass the time parameter in that argument?

File: tests/engines/test_torch.py (hunk 9 above)
Comment (vytautas-a) [reply tests/engines/test_torch.py:103]: good catch, I'll update the test

File: qadence/backends/pyqtorch/convert_ops.py [hunk 12]
```diff
@@ -98,6 +98,71 @@ def extract_parameter(block: ScaleBlock | ParametricBlock, config: Configuration
...
+    """
+
+    # Function to replace the underscore with a dot
+    def underscore_to_dot(match: re.Match) -> Any:
+        return match.group(0).replace("_", ".")
+
+    # Compile the regular expression
+    pattern = re.compile(float_with_underscore_pattern, re.VERBOSE)
+
+    return pattern.sub(underscore_to_dot, s)
```
[the remark this answers is not in the document]
Comment (vytautas-a) [reply qadence/backends/pyqtorch/convert_ops.py:120]: Ok, I'll add it.

**Comment by vytautas-a** [comment]
> Thanks @vytautas-a :)
> 
> Few minor things and then it should be good to go.
> 
> One thing, I see that the docs for the time-dependent HamEvo reads:
> 
> ```
> # create parameterized HamEvo block
> hamevo = HamEvo(generator_td, 0.0, duration=duration)
> ```
> 
> Is passing `0.0` correct there? I thought the idea was to pass the time parameter in that argument?

Yes, the docs have to updated.

File: tests/engines/test_torch.py (hunk 9 above)
Comment (vytautas-a) [reply tests/engines/test_torch.py:103]: It turns out that I already removed the `orig_param_values` key, thus the test is actually correct.

Comment (vytautas-a) [reply tests/engines/test_torch.py:103]: I just removed the unnecessary code from these tests.

File: qadence/backends/pyqtorch/convert_ops.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (vytautas-a) [reply qadence/backends/pyqtorch/convert_ops.py]: done

**Review by jpmoutinho — APPROVED** [review]
LGTM!
